python/lib/core/dmod/core/common/collection.py

```diff
@@ -0,0 +1,46 @@
+"""
+Defines specialized collections that aren't built into any of the first or third party libraries
+"""
+import typing
+
+from collections.abc import Collection
+from typing import Iterator
+
+_T = typing.TypeVar("_T")
+
+
+class Bag(Collection[_T]):
```

Comment:
Well count and remove are easy to add (just did). Tempted to add a `pick` or `choose` function, but that's unnecessary. I avoided Dict[_T, int] to fulfill the requirement for _T to _not_ have to be hashable. Want to use a dict? Key's gotta be hashable. If you need to really reinforce "this is unordered, non-unique, and values don't have to be hashed", a `list` is out the window and a `set` just won't work. This helps in comparisons where you're checking "Is this a sequence? Does the order AND contents match? Yes? Do this. Is this just an iterable? Are all the contents present? Yes? Do this." I really need to do an `__eq__` implementation on it since bags are just being fed to functions that determine the equality themselves.

Reply:
> Well count and remove are easy to add (just did). Tempted to add a pick or choose function, but that's unnecessary.

Nice! I guess that now means that values have to be `eq` or partial `eq`. 

> Want to use a dict? Key's gotta be hashable.

Good catch, I was not thinking about that. A `list` it is then!

>  I really need to do an __eq__ implementation on it since bags are just being fed to functions that determine the equality themselves.

I guess this implies that values must also be orderable? Right?

Reply:
It'd be membership - does Bag A have the same contents as Bag B? So `Bag([1, 3, [4, 5, 6], {"a" 7, "b": 8}]) == Bag([{"b": 8, "a": 7}, 3, [4, 5, 6], 1])`

Reply:
Right, but to have a membership test, I think that values have to be partial `eq`. For example, these two bags should not be equal `Bag([math.nan]) != Bag([math.nan])`.